### PR TITLE
Add default attention to gpt and generate example

### PR DESCRIPTION
**Pull request by smallv0221**
### PR types

Others
### PR changes

Models
### Description

Add default attention to gpt and generate example

### CODE REVIEW

File: paddlenlp/transformers/gpt/modeling.py [hunk 1]
```diff
@@ -195,11 +197,14 @@ def forward(self,
...
+        product = product * causal_mask
+        mask_score = (causal_mask - 1.0) * 10000.0
+        product = product + mask_score
         if attn_mask is not None:
-            # product = product + attn_mask
-            product = product * attn_mask
-            mask_score = (attn_mask - 1.0) * 10000.0
-            product = product + mask_score
+            product = product + attn_mask
+
```
Comment (ZHUI) [inline paddlenlp/transformers/gpt/modeling.py:202]: 1. Why do we need `attn_mask` and `causal_mask` in the same time?
2. The `causal_mask` should be the sequence length, but it is initialized with `embed_dim`.  where `key_length` > `embed_dim`, there may be serious bugs.

**Review by ZHUI — CHANGES_REQUESTED** [review]

File: paddlenlp/transformers/gpt/modeling.py [hunk 2]
```diff
@@ -685,9 +713,11 @@ def forward(self,
                                                          input_ids)
         embedding_output = self.embeddings(
             input_ids=input_ids, position_ids=position_ids)
-
+        causal_mask = self.bias[:paddle.shape(input_ids)[-1], :paddle.shape(
+            input_ids)[-1]]
```
Comment (guoshengCS) [inline paddlenlp/transformers/gpt/modeling.py:704]: 建议causal_mask在这里和传进来的attention_mask使用相加或者与或合起来，后面只传attention_mask，简单一些，在每层的attention中也减少一个mask的处理更快一点

Comment (guoshengCS) [reply paddlenlp/transformers/gpt/modeling.py:704]: 刚才和 @smallv0221  讨论，这个也依赖于attention_mask的统一。当前对各模型、tokenzier、生成API对mask的要求没有统一起来，有的是0-1，有的是1-0，有的是0, -INF，相对完备一些的是之前paddle.nn.Transformer中的处理，这个也见之前的说明 https://github.com/PaddlePaddle/PaddleNLP/pull/311#discussion_r621985230

File: paddlenlp/transformers/gpt/tokenizer.py [hunk 3]
```diff
@@ -98,9 +98,9 @@ def __init__(
             model_file,
             max_len=512,
             unk_token='<unk>',
-            bod_token='<bod>',
-            eod_token='<eod>',
-            stop_token='\u2583',  # The token of newline.
+            bos_token='<bod>',
+            eos_token='<eod>',
+            eol_token='\u2583',  # The token of newline.
```
Comment (guoshengCS) [inline paddlenlp/transformers/gpt/tokenizer.py:103]: 这种换参数名的方式可能会导致一些兼容性的风险，此前用户本地存的配置可能会无法使用。这种情况是要保留原有的，还是咱们增强PretrainedTokenizer/PretrainedTokenizer提供对参数版本升级的支持。另外包括bos上的模型，咱们是否也需要一个版本升级的支持，比如之前有一些预训练模型没有head部分的参数。这个也请  @ZeyuChen 看下

File: paddlenlp/transformers/gpt/modeling.py (hunk 2 above)
Comment (ZHUI) [reply paddlenlp/transformers/gpt/modeling.py:704]: 这里 GPT mask 采用bool mask方式，分布式的同学反馈有性能问题。经查，`mask_score = (causal_mask - 1.0) * 10000.0` 的一些写法有性能问题。建议 causal_mask，attention_mask 统一改成 -INF的mask，统一起来。 @smallv0221

File: examples/language_model/gpt/run_pretrain_static.py [hunk 4]
```diff
@@ -242,7 +244,7 @@ def do_train(args):
                     GPTPretrainingCriterion)()
                 loss = criterion(preds, labels, loss_mask)
 
-            # Create the learning_rate sheduler and optimizer
+            # Create the learning_rate sheduler and optimizergit
```
Comment (ZHUI) [inline examples/language_model/gpt/run_pretrain_static.py]: typos

File: paddlenlp/transformers/gpt/modeling.py [hunk 5]
```diff
@@ -267,25 +267,29 @@ def forward(self,
...
                     new_caches.append(new_cache)
                 else:
-                    output = mod(output,
-                                 memory,
-                                 tgt_mask=tgt_mask,
-                                 use_cache=use_cache,
-                                 cache=cache)
+                    output = mod(
+                        output,
+                        causal_mask,
```
Comment (ZHUI) [inline paddlenlp/transformers/gpt/modeling.py]: typo ? `causal_mask`

File: paddlenlp/transformers/gpt/modeling.py [hunk 6]
```diff
@@ -602,10 +619,18 @@ def __init__(self,
...
                  initializer_range=0.02,
                  pad_token_id=0,
+                 eos_token_id=7,
+                 bos_token_id=0,
+                 eol_token_id=3,
                  topo=None):
         super(GPTModel, self).__init__()
         self.pad_token_id = pad_token_id
         self.initializer_range = initializer_range
+        self.register_buffer("bias",
```
Comment (ZHUI) [inline paddlenlp/transformers/gpt/modeling.py]: 是否验证导出 infer 功能正常？

File: paddlenlp/transformers/gpt/modeling.py [hunk 7]
```diff
@@ -681,11 +698,16 @@ def forward(self,
...
-                                                         input_ids)
+            position_ids = paddle.fluid.layers.expand_as(
+                position_ids, input_ids)
         embedding_output = self.embeddings(
             input_ids=input_ids, position_ids=position_ids)
-
+        causal_mask = self.bias[:paddle.shape(input_ids)[-1], :paddle.shape(
+            input_ids)[-1]]
+        if attention_mask is not None:
+            attention_mask = attention_mask + causal_mask
```
Comment (ZHUI) [inline paddlenlp/transformers/gpt/modeling.py:706]: 此处的如果传进来 `attention_mask ` 是 `causal_mask`， 相当于还是做了一次加法对吧？

File: .pre-commit-config.yaml [hunk 8]
```diff
@@ -1,10 +1,11 @@
+repos:
 -   repo: https://github.com/PaddlePaddle/mirrors-yapf.git
-    sha: 0d79c0c469bab64f7229c9aca2b1186ef47f0e37
+    rev: v0.16.2
```
Comment (ZHUI) [inline .pre-commit-config.yaml]: 此处的 yapf 版本修改，是否会导致大面积的代码格式化变动？

**Review by ZHUI — APPROVED** [review]
LGTM
